precompiles/randomness/src/lib.rs

```diff
@@ -0,0 +1,466 @@
+		return Err(revert("Gas limit must exceed overhead call cost"));
+	}
+	Ok(())
+}
+
+/// Subcall to provide randomness
+/// caller must call `ensure_can_provide_randomness` before calling this function
+fn provide_randomness(
+	handle: &mut impl PrecompileHandle,
+	request_id: u64,
+	gas_limit: u64,
+	contract: H160,
+	randomness: Vec<H256>,
+) -> EvmResult<()> {
+	let (reason, _) = handle.call(
+		contract,
+		None,
+		// callback function selector: keccak256("rawFulfillRandomWords(uint256,uint256[])")
+		EvmDataWriter::new_with_selector(0x1fe543e3_u32)
+			.write(request_id)
+			.write(randomness)
+			.build(),
+		Some(gas_limit),
+		false,
+		&Context {
+			caller: handle.context().address,
+			address: contract,
+			apparent_value: U256::zero(),
+		},
+	);
+	// Logs
+	// We reserved enough gas so this should not OOG.
+	match reason {
+		ExitReason::Revert(_) | ExitReason::Error(_) => {
+			let log = log_fulfillment_failed(handle.code_address());
+			handle.record_log_costs(&[&log])?;
+			log.record(handle)?
+		}
+		ExitReason::Succeed(_) => {
+			let log = log_fulfillment_succeeded(handle.code_address());
+			handle.record_log_costs(&[&log])?;
+			log.record(handle)?
+		}
+		_ => (),
+	}
+	Ok(())
+}
+
+/// A precompile to wrap the functionality from pallet-randomness
+pub struct RandomnessWrapper<Runtime>(PhantomData<Runtime>);
+
+impl<Runtime> Precompile for RandomnessWrapper<Runtime>
+where
+	Runtime: pallet_randomness::Config + pallet_evm::Config + pallet_base_fee::Config,
+	<Runtime as frame_system::Config>::BlockNumber: From<u32>,
+	BalanceOf<Runtime>: TryFrom<U256> + Into<U256>,
+{
+	fn execute(handle: &mut impl PrecompileHandle) -> EvmResult<PrecompileOutput> {
+		log::trace!(target: "randomness-precompile", "In randomness wrapper");
+
+		let selector = handle.read_selector()?;
+
+		handle.check_function_modifier(match selector {
+			Action::RelayEpochIndex | Action::GetRequestStatus | Action::GetRequest => {
+				FunctionModifier::View
+			}
+			_ => FunctionModifier::NonPayable,
+		})?;
+
+		match selector {
+			Action::RelayEpochIndex => Self::relay_epoch_index(handle),
+			Action::GetRequestStatus => Self::get_request_status(handle),
+			Action::GetRequest => Self::get_request(handle),
+			Action::RequestRelayBabeEpochRandomWords => Self::request_babe_randomness(handle),
+			Action::RequestLocalVRFRandomWords => Self::request_local_randomness(handle),
+			Action::FulfillRequest => Self::fulfill_request(handle),
+			Action::IncreaseRequestFee => Self::increase_request_fee(handle),
+			Action::PurgeExpiredRequest => Self::execute_request_expiration(handle),
+		}
+	}
+}
+
+impl<Runtime> RandomnessWrapper<Runtime>
+where
+	Runtime: pallet_randomness::Config + pallet_evm::Config + pallet_base_fee::Config,
+	<Runtime as frame_system::Config>::BlockNumber: TryInto<u64> + TryFrom<u64>,
+	BalanceOf<Runtime>: TryFrom<U256> + Into<U256>,
+{
+	fn relay_epoch_index(handle: &mut impl PrecompileHandle) -> EvmResult<PrecompileOutput> {
```

Comment:
How are we expecting this to be used?

It might not be very useful without some additional context (such as how long before the next epoch starts). If this is the case, a typical user would want to query this and some additional info (current block number, expected number of blocks until next epoch, etc.)

So it might make for a better (at least cheaper) API if we provided this extra info, then.

Reply:
It actually isn't useful at all now that we just default to requesting 2 epochs after the epoch in which we made the request. The user doesn't need it.

@crystalin can I remove it

Reply:
Yes it is useful for the smart contract to follow the current relay epoch.
But we can remove it for now and reinclude when we have more demand for it

Reply:
> Yes it is useful for the smart contract to follow the current relay epoch.

Ok, let's just keep it then. 

We can see if there is more demand to see when the epoch will change and add that later.